File: 백준/Silver/11055. 가장 큰 증가하는 부분 수열/가장 큰 증가하는 부분 수열.py
# 10
# 1 100 2 50 60 3 5 6 7 8
# 이 입출력 예제보면 1,2,50,60 이 선택되어 답은 113

# 우선 각 자리마다 구한 배열 dp[1:]은 아래와 같다
# [1, 101, 3, 53, 113, 6, 11, 17, 24, 32]

import sys
input = sys.stdin.readline

N = int(input())
arr = [0] + list(map(int, input().strip().split()))
dp = [0] * (N+1)
dp[1] = arr[1]

for i in range(2, N+1):
    for j in range(i):
        if arr[i] > arr[j]:
            dp[i] = max(dp[i], dp[j]+arr[i])

print(max(dp))

File: 백준/Silver/11055. 가장 큰 증가하는 부분 수열/가장 큰 증가하는 부분 수열.java
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.BufferedReader;

import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        int N = Integer.parseInt(br.readLine());
        int[] arr = new int[N+1];
        int[] dp = new int[N+1];

        st = new StringTokenizer(br.readLine(), " ");
        for (int i = 1; i < N+1; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        // dp
        int max_ = 0;
        for (int i = 1; i < N+1; i++) {
            for (int j = 0; j < i; j++) {
                if (arr[i] > arr[j]) {
                    dp[i] = Math.max(dp[i], dp[j] + arr[i]);
                }
            }
            max_ = Math.max(max_, dp[i]);
        }
        System.out.println(max_);



    }
}